### PR TITLE
Pfappserver Updates

### LINKED ISSUE #3618: API-based Web Admin: remove dependency to moment.js
[moment.js](https://momentjs.com/) is big and probably not necessary. We should try to replace it by [date-fns](https://github.com/date-fns/date-fns) (already integrated).

See [You don't (may not) need Moment.js](https://github.com/you-dont-need/You-Dont-Need-Momentjs).

### PR DESCRIPTION
**Pull request by satkunas**
# Description
Continue work on the pfappserver

# Impacts
Increased PR count :100: 

# Issue
* Fixes #3618 

# Delete branch after merge
YES

### CODE REVIEW

File: html/pfappserver/root/static.alt/src/components/pfFormChosen.vue [hunk 1]
```diff
@@ -0,0 +1,155 @@
+<template>
+  <b-form-group horizontal :label-cols="(columnLabel) ? labelCols : 0" :label="$t(columnLabel)"
+    :state="isValid()" :invalid-feedback="getInvalidFeedback()" :class="['chosen-element', { 'mb-0': !columnLabel }, { 'is-focus': focus}]">
```
Comment (cgx) [inline html/pfappserver/root/static.alt/src/components/pfFormChosen.vue]: `class` and `v-bind:class` can co-exist, so I propose:

```html
  <b-form-group .. class="chosen-element" :class="{'mb-0': !columnLabel, 'is-focus': focus }">
```

File: html/pfappserver/root/static.alt/src/components/pfFormChosen.vue [hunk 2]
```diff
@@ -0,0 +1,155 @@
...
+  </b-form-group>
+</template>
+
+<script>
+import Multiselect from 'vue-multiselect'
+import 'vue-multiselect/dist/vue-multiselect.min.css'
+import pfMixinValidation from '@/components/pfMixinValidation'
+
+export default {
+  name: 'pf-form-input',
```
Comment (cgx) [inline html/pfappserver/root/static.alt/src/components/pfFormChosen.vue]: *pf-form-chosen*

File: html/pfappserver/root/static.alt/src/components/pfFormChosen.vue [hunk 3]
```diff
@@ -0,0 +1,155 @@
+<template>
+  <b-form-group horizontal :label-cols="(columnLabel) ? labelCols : 0" :label="$t(columnLabel)"
+    :state="isValid()" :invalid-feedback="getInvalidFeedback()" :class="['chosen-element', { 'mb-0': !columnLabel }, { 'is-focus': focus}]">
+    <b-input-group class="input-group-chosen">
```
Comment (cgx) [inline html/pfappserver/root/static.alt/src/components/pfFormChosen.vue]: Where is this class defined?

File: html/pfappserver/root/static.alt/src/components/pfFormDatetime.vue [hunk 4]
```diff
@@ -16,21 +16,25 @@
  *      example :moments="['-1 hours', '1 hours', '1 days', '1 weeks', '1 months', '1 quarters', '1 years']"
  */
  <template>
-  <b-form-group horizontal :label-cols="(columnLabel) ? labelCols : 0" :label="$t(columnLabel)"
-    :state="isValid()" :invalid-feedback="getInvalidFeedback()" :class="{ 'mb-0': !columnLabel }">
-    <b-input-group>
-      <b-input-group-prepend v-if="prependText" is-text>
-        {{ prependText }}
+  <b-form-group horizontal :label-cols="(columnLabel) ? labelCols : 0" :label="$t(columnLabel)" 
+    :state="isValid()" :invalid-feedback="getInvalidFeedback()" :class="['datetime-element', { 'mb-0': !columnLabel }, { 'is-focus': focus}]">
```
Comment (cgx) [inline html/pfappserver/root/static.alt/src/components/pfFormDatetime.vue]: Again, I propose:

```html
<b-form-group .. class="datetime-element" :class="{ 'mb-0': !columnLabel, 'is-focus': focus }">
```

File: html/pfappserver/root/static.alt/src/components/pfFormInput.vue [hunk 5]
```diff
@@ -1,6 +1,6 @@
 <template>
   <b-form-group horizontal :label-cols="(columnLabel) ? labelCols : 0" :label="$t(columnLabel)"
-    :state="isValid()" :invalid-feedback="getInvalidFeedback()" :class="{ 'mb-0': !columnLabel }">
+    :state="isValid()" :invalid-feedback="getInvalidFeedback()" :class="['input-element', { 'mb-0': !columnLabel }]">
```
Comment (cgx) [inline html/pfappserver/root/static.alt/src/components/pfFormInput.vue]:
```html
<b-form-group .. class="input-element" :class="{ 'mb-0': !columnLabel }">
```

File: html/pfappserver/root/static.alt/src/components/pfFormPrefixMultiplier.vue [hunk 6]
```diff
@@ -1,18 +1,23 @@
 <template>
-  <b-form-group horizontal :label-cols="(columnLabel) ? labelCols : 0" :label="$t(columnLabel)"
-    :state="isValid()" :invalid-feedback="getInvalidFeedback()" :class="{ 'mb-0': !columnLabel }">
-    <b-input-group>
-      <b-input-group-prepend v-if="prependText" is-text>
-        {{ prependText }}
+  <b-form-group horizontal :label-cols="(columnLabel) ? labelCols : 0" :label="$t(columnLabel)" 
+    :state="isValid()" :invalid-feedback="getInvalidFeedback()" :class="['prefixmultiplier-element', { 'mb-0': !columnLabel }, { 'is-focus': focus}]">
```
Comment (cgx) [inline html/pfappserver/root/static.alt/src/components/pfFormPrefixMultiplier.vue]:
```html
<b-form-group .. class="prefixmultiplier-element" :class="{ 'mb-0': !columnLabel, 'is-focus': focus }">
```

File: html/pfappserver/root/static.alt/src/components/pfFormPrefixMultiplier.vue [hunk 7]
```diff
@@ -183,20 +189,68 @@ export default {
...
+  created () {
+    if (this.value) {
+      this.realValue = this.value
+    }
+    this.setInputValueFromRealValue()
   }
 }
 </script>
 
 <style lang="scss" scoped>
```
Comment (cgx) [inline html/pfappserver/root/static.alt/src/components/pfFormPrefixMultiplier.vue:211]: Does it really need to be scoped?

File: html/pfappserver/root/static.alt/src/globals/pfValidators.js [hunk 8]
```diff
@@ -161,3 +242,144 @@ export const userNotExists = (value, component) => {
...
+    if (params) {
+      // iterate through all params
+      for (let i = 0; i < params.length; i++) {
+        const [param] = params[i] // destructure
+        if (parent[param].$model === undefined) continue // ignore empty models
+        if (idOfV(parent[param].$model) === id) continue // ignore (self)
+        // iterate through _fieldTypes and substitute
+        _fieldTypes = _fieldTypes.map(fieldType => {
+          // substitute the fieldType with |true| if it exists
+          return (parent[param].$model.type === fieldType) ? true : fieldType
```
Comment (cgx) [inline html/pfappserver/root/static.alt/src/globals/pfValidators.js]: We should return immediately when the condition is true.

File: html/pfappserver/root/static.alt/src/globals/pfValidators.js [hunk 9]
```diff
@@ -161,3 +242,144 @@ export const userNotExists = (value, component) => {
...
```
Comment (cgx) [inline html/pfappserver/root/static.alt/src/globals/pfValidators.js]: We should return immediately when the condition is true.

File: html/pfappserver/root/static.alt/src/views/Users/_components/UsersCreate.vue [hunk 10]
```diff
@@ -80,18 +80,55 @@
...
+                  ]"
+                />
               </b-col>
             </b-form-row>
           </b-form-group>
+
+          <pf-form-sortable-fields
+            sortable
+            v-model="actions"
+            :column-label="$t('Actions')"
```
Comment (cgx) [inline html/pfappserver/root/static.alt/src/views/Users/_components/UsersCreate.vue]: The label will be localized by the component.

File: html/pfappserver/root/static.alt/src/views/Users/_components/UsersCreate.vue [hunk 11]
```diff
@@ -80,18 +80,55 @@
...
+            :invalid-feedback="[
+              { [$t('One or more errors exist.')]: !$v.actions.anyError }
+            ]"
+            @validations="actionsValidations = $event"
+          ></pf-form-sortable-fields>
+
+          <!--
+          <h1>$v</h1>
+          <pre>{{ JSON.stringify($v.actions, null, 2) }}</pre>
+          -->
```
Comment (cgx) [inline html/pfappserver/root/static.alt/src/views/Users/_components/UsersCreate.vue]: Remove comments.

File: html/pfappserver/root/static.alt/src/views/Users/_components/UsersCreate.vue [hunk 12]
```diff
@@ -106,32 +143,46 @@
...
 
 export default {
   name: 'UsersCreate',
   components: {
-    draggable,
+    'pf-form-datetime': pfFormDatetime,
     'pf-form-input': pfFormInput,
+    'pf-form-sortable-fields': pfFormSortableFields,
     'pf-form-textarea': pfFormTextarea,
     'pf-form-toggle': pfFormToggle
```
Comment (cgx) [inline html/pfappserver/root/static.alt/src/views/Users/_components/UsersCreate.vue]: When the component's name is properly defined, you can omit the keys:

```javascript
components: {
    pfFormDatetime,
    pfFormInput,
    pfFormSortableFields,
    pfFormTextarea,
    pfFormToggle
}
```

File: html/pfappserver/root/static.alt/src/utils/bytes.js [hunk 13]
```diff
@@ -0,0 +1,26 @@
+const bytes = {
+  toHuman (bytes, precision=2, abbreviate=false) {
+    if (bytes === 0) return bytes.toString()
```
Comment (cgx) [inline html/pfappserver/root/static.alt/src/utils/bytes.js]: We need to append a space at the end.

File: html/pfappserver/root/static.alt/src/globals/pfValidators.js [hunk 14]
```diff
@@ -161,3 +242,144 @@ export const userNotExists = (value, component) => {
...
+export const limitSiblingFieldTypes = (limit) => {
+  return (0, _common.withParams)({
+    type: 'limitSiblingFieldTypes',
+    limit: limit
+  }, function (value, field) {
+    let count = 0
+    const { id, parent, params } = idParentParamsFromV(this.$v, field)
+    if (params) {
+      // iterate through all params
+      for (let i = 0; i < params.length; i++) {
```
Comment (cgx) [inline html/pfappserver/root/static.alt/src/globals/pfValidators.js:265]: We could add a condition to return sooner if the count is bigger than the limit.

File: html/pfappserver/root/static.alt/src/views/Users/_components/UsersCreate.vue [hunk 15]
```diff
@@ -144,15 +195,143 @@ export default {
...
         pid: '',
         email: '',
-        password: '',
-        chosen: []
+        password: ''
       },
-      valid_from: new Date()
+      valid_from: null,
+      expiration: null,
+      actions: JSON.parse('[ { "type": "set_bandwidth_balance", "value": 64 }, { "type": "set_access_duration", "value": "" }, { "type": "set_access_level", "value": "User Manager" }, { "type": "mark_as_sponsor", "value": null }, { "type": "set_role", "value": "1" }, { "type": "set_tenant_id", "value": "1" }, { "type": "set_time_balance", "value": "12h" }, { "type": "set_unreg_date", "value": "2019-08-05 01:00:00" } ]'),
```
Comment (cgx) [inline html/pfappserver/root/static.alt/src/views/Users/_components/UsersCreate.vue]: Values for testing?

**Review by cgx — CHANGES_REQUESTED** [review]

File: html/pfappserver/root/static.alt/src/views/Users/_components/UsersCreate.vue (hunk 15 above)
Comment (satkunas) [reply html/pfappserver/root/static.alt/src/views/Users/_components/UsersCreate.vue]: Yes, exact

File: html/pfappserver/root/static.alt/src/utils/bytes.js (hunk 13 above)
Comment (satkunas) [reply html/pfappserver/root/static.alt/src/utils/bytes.js]: Nice catch, good :eye:

File: html/pfappserver/root/static.alt/src/components/pfFormPrefixMultiplier.vue (hunk 7 above)
Comment (satkunas) [reply html/pfappserver/root/static.alt/src/components/pfFormPrefixMultiplier.vue:211]: Yes, normally the inputValue/value is used directly in the form input element. In this component it utilizes a 2nd alternate model that differs from what's displayed in the form input (depending on the selected `prefix`).

File: html/pfappserver/root/static.alt/src/components/pfFormChosen.vue [hunk 16]
```diff
@@ -1,7 +1,8 @@
 <template>
   <b-form-group horizontal :label-cols="(columnLabel) ? labelCols : 0" :label="$t(columnLabel)"
-    :state="isValid()" :invalid-feedback="getInvalidFeedback()" :class="['chosen-element', { 'mb-0': !columnLabel }, { 'is-focus': focus}]">
-    <b-input-group class="input-group-chosen">
+    :state="isValid()" :invalid-feedback="getInvalidFeedback()" 
+    class="chosen-element" :class="[{ 'mb-0': !columnLabel }, { 'is-focus': focus}]">
```
Comment (cgx) [inline html/pfappserver/root/static.alt/src/components/pfFormChosen.vue]: You can simplify to a single object literal: `:class="{ 'mb-0': !columnLabel, 'is-focus': focus}"`

File: html/pfappserver/root/static.alt/src/components/pfFormDatetime.vue [hunk 17]
```diff
@@ -17,7 +17,8 @@
  */
  <template>
   <b-form-group horizontal :label-cols="(columnLabel) ? labelCols : 0" :label="$t(columnLabel)" 
-    :state="isValid()" :invalid-feedback="getInvalidFeedback()" :class="['datetime-element', { 'mb-0': !columnLabel }, { 'is-focus': focus}]">
+    :state="isValid()" :invalid-feedback="getInvalidFeedback()" 
+    class="datetime-element" :class="[{ 'mb-0': !columnLabel }, { 'is-focus': focus}]">
```
Comment (cgx) [inline html/pfappserver/root/static.alt/src/components/pfFormDatetime.vue]: You can simplify to a single object literal: `:class="{ 'mb-0': !columnLabel, 'is-focus': focus }"`

File: html/pfappserver/root/static.alt/src/components/pfFormPrefixMultiplier.vue [hunk 18]
```diff
@@ -1,6 +1,7 @@
 <template>
   <b-form-group horizontal :label-cols="(columnLabel) ? labelCols : 0" :label="$t(columnLabel)" 
-    :state="isValid()" :invalid-feedback="getInvalidFeedback()" :class="['prefixmultiplier-element', { 'mb-0': !columnLabel }, { 'is-focus': focus}]">
+    :state="isValid()" :invalid-feedback="getInvalidFeedback()" 
+    class="prefixmultiplier-element" :class="[{ 'mb-0': !columnLabel }, { 'is-focus': focus}]">
```
Comment (cgx) [inline html/pfappserver/root/static.alt/src/components/pfFormPrefixMultiplier.vue]: You can simplify to a single object literal: `:class="{ 'mb-0': !columnLabel, 'is-focus': focus }"`

File: html/pfappserver/root/static.alt/src/components/pfFormSortableFields.vue [hunk 19]
```diff
@@ -77,7 +77,8 @@
 -->
 <template>
   <b-form-group horizontal :label-cols="(columnLabel) ? labelCols : 0" :label="$t(columnLabel)"
-    :state="isValid()" :invalid-feedback="getInvalidFeedback()" :class="['sortablefields-element', { 'is-focus': drag }, { 'mb-0': !columnLabel }]"
+    :state="isValid()" :invalid-feedback="getInvalidFeedback()" 
+    class="sortablefields-element" :class="[{ 'is-focus': drag }, { 'mb-0': !columnLabel }]"
```
Comment (cgx) [inline html/pfappserver/root/static.alt/src/components/pfFormSortableFields.vue]: You can simplify to a single object literal: `:class="{ 'is-focus': drag, 'mb-0': !columnLabel }"`

**Review by cgx — APPROVED** [review]

**Pull request merged**